pandas/tests/frame/indexing/test_setitem.py

```diff
@@ -338,6 +338,16 @@ def test_setitem_bool_with_numeric_index(self, dtype):
 
         tm.assert_index_equal(df.columns, expected_cols)
 
+    @pytest.mark.parametrize("indexer", ["B", ["B"]])
+    def test_setitem_frame_length_0_str_key(self, indexer):
+        # GH#38831
+        df = DataFrame(columns=["A", "B"])
+        other = DataFrame({"B": [1, 2]})
+        df[indexer] = other
```

Comment:
do we need/want to check any of the other indexers?

Reply:
Don't know if this would be helpful, the string-DataFrame case is a special case while indexer in [list, Index, Series, array] is all the same (which was working previously)

Reply:
hmm, can you check `df.loc[:, indexer]` ? should be the same result i think.

Reply:
No unfortunately not, in case of ``indexer=["B"]``  we align, which leads to an empty result. This is losely related to our dicussion in #30439 @jbrockmendel 

When ``indexer="B"`` this works as the case above